### DISCUSSION TITLE
Is it possible to connect NoMali GPU device to the ARM SMMU?

### DISCUSSION THREAD

**Discussion by sammy17** [gem5/gem5#1003] [Q&A] [answered]
As the title states, I want to know where it's possible to connect the NoMali GPU device that's already available in gem5 to the ARM SMMU. 
In a more general sense, I guess the question is whether it's possible to connect device types other than PCiDevices to the SMMU. 
I am not restricted to NoMali GPU, I am trying to get any GPU device model working with ARM SMMU. 

I noticed that AMD GPU is available as a PciDevice but it is restricted to x86. 

I really appreciate any help on this. Thank you in advance.

**Comment by giactra** [comment, marked as answer]
Absolutely, the SMMU is not required to be used by the PCIe root complex only, and any device could be connected to it. Have a look at the System Placement (2.9) chapter in https://developer.arm.com/documentation/ihi0070/latest to get a visual example..

**Reply by ferry-hhh** [reply]
Hi @giactra ,

I am new to gem5, and I have a question regarding SMMU. I see that SMMU is implemented for the Arm architecture. I would like to know if it is possible to use it on the x86 architecture (as an IOMMU)? Alternatively, how do IDE devices and AMD GPU devices on x86 architecture perform virtual to physical address translation before DMA?

Any guidance or insights would be greatly appreciated. Thank you in advance for your time and assistance!

**Comment by sammy17** [comment]
Thanks @giactra, the figure was helpful. Just to confirm, in gem5, NoMali device should be connected through two IO interconnects (one for incoming transactions and one for outgoing)?
Also, since the NoMali GPU device is defined using the PioDevice class, how is the sid (stream ID) of the device generated?

**Comment by giactra** [comment]
> Thanks @giactra, the figure was helpful. Just to confirm, in gem5, NoMali device should be connected through two IO interconnects (one for incoming transactions and one for outgoing)?

You don't have to. I have no experience with NoMali, but you can connect any DmaDevice directly to the SMMU through the dma port and connect the programmable interface (the pio port) to the iobus.

> Also, since the NoMali GPU device is defined using the PioDevice class, how is the sid (stream ID) of the device generated?

https://github.com/gem5/gem5/blob/stable/src/dev/Device.py#L103

Once you set sid, it will be visible to the DTB when you autogenerate. Otherwise you would have to amend the FDT manually
